keychain_haiku.cpp

```diff
@@ -0,0 +1,129 @@
+/******************************************************************************
+ *   Copyright (C) 2011-2015 Frank Osterfeld <frank.osterfeld@gmail.com>      *
+ *                                                                            *
+ * This program is distributed in the hope that it will be useful, but        *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY *
+ * or FITNESS FOR A PARTICULAR PURPOSE. For licensing and distribution        *
+ * details, check the accompanying file 'COPYING'.                            *
+ *****************************************************************************/
+#include "keychain_p.h"
+
+#include <KeyStore.h>
+#include <QDebug>
+
+using namespace QKeychain;
+
+static QString strForStatus( status_t os ) {
+    const char * const buf = strerror(os) ;
+    return QObject::tr( "error 0x%1: %2" )
+            .arg( os, 8, 16 ).arg( QString::fromUtf8( buf, strlen( buf ) ) );
+}
+
+void ReadPasswordJobPrivate::scheduledStart()
+{
+    QString errorString;
+    Error error = NoError;
+    BKeyStore keyStore;
+    BPasswordKey password;
+
+    status_t result = keyStore.GetKey(B_KEY_TYPE_PASSWORD,
+                                      q->service().toUtf8().constData(),
+                                      q->key().toUtf8().constData(),
+                                      false, password);
+
+    data = QByteArray((char*)password.Data(), password.DataLength());
+
+    switch ( result ) {
+    case B_OK:
+        q->emitFinished();
+        return;
+    case B_ENTRY_NOT_FOUND:
+        errorString = tr("Password not found");
+        error = EntryNotFound;
+        break;
+    default:
+        errorString = strForStatus( result );
+        error = OtherError;
+        break;
+    }
+
+    q->emitFinishedWithError( error, errorString );
+}
+
+void WritePasswordJobPrivate::scheduledStart()
+{
+    QString errorString;
+    Error error = NoError;
+    BKeyStore keyStore;
+    BPasswordKey password(data.constData(),
+        B_KEY_PURPOSE_GENERIC,
+        q->service().toUtf8().constData(),
+        q->key().toUtf8().constData());
+    status_t result;
+
+    // XXX: is it a string or a blob??
```

Comment:
That depends on JobPrivate::mode

Reply:
I don't know our API much on this topic, and it hasn't been tested much yet, but I think the BPasswordKey class expects a string, while the base class is more flexible. Maybe I should just use the base class and set the fields separately.